User manip test 1

bgoncalv opened the pull request:
Created a class to allow execution of multiple commands using same docker container.
Created test case for user creation, modification and deletion.

bgoncalv committed:
refactor and separate docker command handling

bgoncalv committed:
Added class BaseRuntimeSmokeTest to allow run multiple commands on same docker container and added test case to test user  creation, modification and deletion

bgoncalv committed:
fixed some mis-spelled words and code style

mmathesius reviewed (changes requested):
I like the direction you took with the enhancements and new test!

However, I do have some concerns about error/exception handling. For the BaseRuntimeRunCmd class, I see exceptions are caught and error codes returned instead. However, for certain errors, we should probably raise an exception gets handled in the tests so we don't continue when there's a problem in the environment.

In particular, the return code from calls to _rm_br_container() (lines #42 and #83) ignored, but shouldn't be.

I also see an inconsistency in what gets returned from _run_br_container_background(). At line #56, it's returning a code and output rather than a container ID (or None).

Regarding testUserManipulation(), the whole test should just be failed right away if any of the individual commands fail.

I also think there's a problem with the test commands that begin with "su - usertest && ..." at lines #184 and #189. If the intent is that the part of the command after the "&&" is running as user "usertest", that isn't the case. The part after "&&" will only run as root (since that's the user for docker commands) only if "su - usertest" succeeds. Are those commands just intended to be placeholders because the "su" command is currently missing?

Finally, when I attempt to run "smoke.py", the testUserManipulation test fails because of the missing "su" command. And, unfortunately, we don't want to integrate a new test that fails. We really need to wait until the docker image includes "su" before merging this PR.

bgoncalv committed:
BaseRuntimeRunCmd() - raise AssertionError exception if fails to create or remove container.
testUserManipulation() - fixed error when running 'su' and abort execution after first error.

mmathesius commented:
I'm happy with the changes and have merged your PR.

On a future update, however, there are a couple small things I'd like to see improved:

1. With _rm_br_container now fully handling exceptions, it no longer needs to return a value. It gets ignored everywhere it's called anyhow.

2. Rather than relying upon _del_ to cleanup the docker container, but it would be best to clean it up explicitly--perhaps by calling `run_cmd.run_docker_cmd("true", rm_container=True)`. In any case, when you merge in the latest upstream changes, you'll find that teardown.py will also be taking care of it just in case.
